File: sorting/quicksort.hs
-- @Author: krocki
-- @Date:   2016-12-16 18:29:38
-- @Last Modified by:   krocki
-- @Last Modified time: 2016-12-16 20:07:58

import System.Random
import Data.List

main = do
    seed <- newStdGen
    let unsorted = randomList (100, 1, 1000) seed
        sorted = quicksort unsorted
    putStrLn $ "Before sorting: \n" ++ show unsorted
    putStrLn $ "After sorting: \n" ++ show sorted

randomList :: (Int, Int, Int) -> StdGen -> [Int]
randomList (n, min, max) gen = take n (randomRs (min,max) gen)

-- let's choose 1st element to be the pivot (x)
-- we put this pivot in the middle, all elements < x to the left,
-- all elements >= x to the right and process left and right
-- recursively until []

quicksort:: (Ord a) => [a] -> [a]
quicksort [] = []
quicksort (x:xs) = quicksort left  ++ [x] ++ quicksort right
                   where left = [e | e <- xs, e < x]
                         right = [e | e <- xs, e >= x]

File: sorting/quicksort.c
/*
* @Author: krocki
* @Date:   2016-12-19 10:16:33
* @Last Modified by:   krocki
* @Last Modified time: 2017-01-17 21:18:58
*/

#include <stdio.h>

// arr = [3, 4, 5, 1, 2, 4]
#define SWAP(a, b) do { a ^= b; b ^= a; a ^= b; } while(0)

int partition(int* arr, int low, int high) {

	int pivot = arr[low];

	printf("pivot = %d\n", pivot);

	int i = low + 1;
	int j = high;

	while (i < j) {

		printf("i = %d, j = %d, a[i] = %d, a[j] = %d\n", i, j, arr[i], arr[j]);
		while (arr[i] <= pivot) i++;
		while (arr[j] > pivot) j--;

		if (i < j) SWAP(arr[i], arr[j]);

	}

	printf("i = %d, j = %d, a[i] = %d, a[j] = %d\n", i, j, arr[i], arr[j]);

	arr[0] = arr[j];
	arr[j] = pivot;

	return j;

}

void quicksort(int* arr, int low, int high) {

	if (low < high) {

		int i = partition(arr, low, high);
		printf("low = %d, high = %d, i = %d\n", low, high, i);

		// TODO:
	}
}

void print_array(int* arr, int n) {

	for (int i = 0; i < n; i++)
		printf("%3d", arr[i]);

	printf("\n");

}

int main() {

	int n = 10;
	int arr[10] = { 4, 1, 5, 10, 2, 6, 3, 7, 8, 9};

	print_array(arr, n);
	quicksort(arr, 0, n - 1);
	print_array(arr, n);
}

